Nanopore DNA sequencing is portable and relatively cheap, allowing real time sequencing in the field.  We see the potential to use nanopore sequencing as an accessible educational experience. With a clear pipeline that Just Works(TM), a citizen scientist could do WIMP (What's in my pot?) analysis on their own samples without the need for any external tools.  Undergrad or high school students could follow the steps of the pipeline to learn about the basics of genome assembly.

# Why Nanopore is amazing
* Long reads, cheap, portable
* Comparison to the current standard (Illumina)
* Used for detecting ebola/zika?  Sent to space
* Sequencing in the jungle (tweet below!)
* Idea we care about: you should be able to take a random sample of stuff (ocean water?  dirt?) and sequence it cheaply and easily to find out what's there

To see a video of nanopore sequencing in the jungle, click on the below block of code and click the "Run" button at the top of this page.  You can see on the left side of the code block that it's been run x number of times ("In [x]").  You'll see a star while the code is running ("In [*]") and once it's done, you'll be able to see the video.

In [ ]:
%%html
<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">Welcome to my laboratory :)<br><br>Sequencing long ribosomal cluster from plants, insects &amp; fungi in real-time in the Amazon rainforest. Within a few mins of <a href="https://twitter.com/nanopore?ref_src=twsrc%5Etfw">@nanopore</a> data generated, performed BLAST &amp; got correct hits! Dual indexing looks great for pooling many samples<a href="https://twitter.com/hashtag/junglegenomics?src=hash&amp;ref_src=twsrc%5Etfw">#junglegenomics</a> <a href="https://t.co/UQVjYfmU8U">pic.twitter.com/UQVjYfmU8U</a></p>&mdash; Aaron Pomerantz (@AaronPomerantz) <a href="https://twitter.com/AaronPomerantz/status/980873273348038656?ref_src=twsrc%5Etfw">April 2, 2018</a></blockquote>
<script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

# How Nanopore actually works
Insert a good paragraph description here:

My current understanding is just like, DNA strand pulled through a pore, current across pore changes depending on the nucleotide, you end up with an electrical signal in fast5 format that can be converted to a fasta file (could be fastq but no one trusts quality scores?)

## Poretools
The raw data we get from nanopore sequencing is in the fast5 format.  This is just a series of current values that were read across the pore as the DNA strand passed through it.

We are going to start by looking at this fast5 data, containing current values, and converting it to a fasta file that contains nucleotides. 

This poretools tutorial is adapted from here: http://nbviewer.jupyter.org/github/arq5x/poretools/blob/master/poretools/ipynb/test_run_report.ipynb

First we're going to find our fast5 files.  Our sample fast5 file is in the "data" folder, so we set the variable dataDirectory to "data/".  If you are using your own data, change dataDirectory to the path to your .fast5 files.

In [2]:
# dataDirectory is the path to our fast5 file.
# If you are using your own data, change dataDirectory to the path to your .fast5 files.
dataDirectory = 'data/'

# Print the number of fast5 files in the dataDirectory.
# Click the "Run" button at the top of this page to run this code.
!find $dataDirectory -maxdepth 1 -name "*.fast5" | wc -l

2


Poretools has a number of command line options.  Let's start with the stats command, which will give us some basic statistics about our reads.

In [3]:
# The -q option stops poretools outputting any warning messages.
!poretools stats -q $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	6
total base pairs	25217
mean	4202.83
median	4205
min	2940
max	5826
N25	5079
N50	5011
N75	3399


Our sample data has 6 reads and 25,217 base pairs. (Anything else of interest to say about this info?)

We have 3 reads per fast5 because forward, reverse, and two-directional reads are all counted separately. (Is this correct?) We can see the stats about each of these types of reads using the below code.

In [4]:
# Look at stats for forward strands
!poretools stats -q --type fwd $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	2
total base pairs	8019
mean	4009.50
median	4009
min	2940
max	5079
N25	5079
N50	5079
N75	2940


In [5]:
# Look at stats for reverse strands
!poretools stats -q --type rev $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	2
total base pairs	7973
mean	3986.50
median	3986
min	2962
max	5011
N25	5011
N50	5011
N75	2962


In [6]:
# Look at two-directional reads
!poretools stats -q --type 2D $dataDirectory

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
total reads	2
total base pairs	9225
mean	4612.50
median	4612
min	3399
max	5826
N25	5826
N50	5826
N75	3399


Hopefully we are going to add how to make a squiggle plot here at some point, that shows the current changing and gives a good idea of what signal actually looks like.

In [ ]:
# Add squiggle plot here!!!

Now we are going to convert our fast5 file to fasta.  Fasta is a common format for storing DNA sequences.  The below code will take each of the fast5 files in dataDirectory, create a fasta file of that sequence, and store it in a folder called fastaOutput.

In [7]:
# Make a folder to store our fasta files in.
!mkdir fastaOutput

# Convert our fast5 files to fasta.
!poretools fasta $dataDirectory > fastaOutput/outputPoretoolData.fasta

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


We can look at the first few lines of this fasta file to see what's in it.  Each of the sequences has a line containing ">" and then a unique identfier, followed by a line containing the nucleotide sequence.

In [9]:
# This will show us the first 200 characters of the first two lines of our file.
# We don't want to look at the whole sequence because it's going to be really long!
!cut -c -200 fastaOutput/outputPoretoolData.fasta | head -2

>b233f432-7786-4b0b-8b2d-03c2e168a45b_Basecall_2D_2d CPHG_CNU4299G4G_louse_library_2016_3_4_3507_1_ch120_read240_strand data/2016_3_4_3507_1_ch120_read240_strand.fast5
GAAATTGCTCCCGCTCTCAGTTCTGCTTTAACAGATAAATTAATAACATATCAATAAAGCATCAAAATCACGTGATTGGAACGCCGTACTTCGAAGAGGAGGATGGAGACGAGGATGGGAGCAGAGGGGAGGATGTGCACTTCTCCCCACGTCAGTTGGGATTCGAAGGAAGTTTGCGGCTTGTTTTAGAGTGGAGGACA


This fasta file containing our raw reads will be the input to the next steps in our pipeline.

In [8]:
!poretools winner $dataDirectory > winner.fasta

/usr/local/lib/python2.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters


# Alignment and assembly

(I don't really know how to explain this so I'll put Brandon's explanation in here)

Minimap2, we take our fasta file of raw reads and we try to align all of the reads to each other (all vs. all).  In order to do this we tell minimap2 to compare our fasta file to itself.  We write this to another file that contains the (alignment?) called overlap.paf.

Because real fast5 files are humongous (very high frequency resolution), we didn't use a real fast5 sample for poretools.  For this step, we want real sample data, so we're actually going to use a different fasta file.  We're going to wget this and then run minimap2 on it.


In [ ]:
!wget https://nanopore.s3.climb.ac.uk/MAP006-1_2D_pass.fasta

Now we're going to tell minimap2 to try to align these reads to each other and write the overlaps to a file called overlap.paf.

In [13]:
!minimap2 -x ava-ont -t 1 MAP006-1_2D_pass.fasta MAP006-1_2D_pass.fasta.2 > overlap.paf

[M::mm_idx_gen::11.601*0.97] collected minimizers
[M::mm_idx_gen::17.437*0.98] sorted minimizers
[M::main::17.437*0.98] loaded/built the index for 25483 target sequence(s)
[M::mm_mapopt_update::18.078*0.98] mid_occ = 52
[M::mm_idx_stat] kmer size: 15; skip: 5; is_hpc: 0; #seq: 25483
[M::mm_idx_stat::18.480*0.98] distinct minimizers: 44056556 (79.15% are singletons); average occurrences: 1.918; average spacing: 2.942
[M::worker_pipeline::214.757*0.93] mapped 25483 sequences
[M::main] Version: 2.10-r763-dirty
[M::main] CMD: minimap2 -x ava-ont -t 1 MAP006-1_2D_pass.fasta MAP006-1_2D_pass.fasta.2
[M::main] Real time: 214.785 sec; CPU: 198.730 sec


Miniasm, we take these overlapping reads and assemble them into contigs.  Need the raw reads from poretools and the overlap file from minimap2 (I don't really understand what this means so once again gonna use a paragraph from brandon).  (because this outputs a .gfa file, we're piping the output into awk and using that to convert it to fasta format, then writing it to asm.fa)

A contig is a combination of several shorter reads that overlap to make one continuous sequence.  This "contiguous" segment of DNA is thought to be a true representation of a piece of the organism's genome.

In [ ]:
!miniasm -f fastaOutput/outputPoretoolData.fasta overlap.paf | awk '/^S/{print ">"$2"\n"$3}' | fold > asm.fa

In [ ]:
# Retrieve genome sequences for each species
!wget ftp://ftp.ensemblgenomes.org/pub/bacteria/release-33/fasta/bacteria_9_collection/escherichia_coli_0_1304/dna/Escherichia_coli_0_1304.GCA_000303235_1.dna.nonchromosomal.fa.gz
!wget ftp://ftp.ensemblgenomes.org/pub/release-38/bacteria//fasta/bacteria_15_collection/_clostridium_asparagiforme_dsm_15981/dna/_clostridium_asparagiforme_dsm_15981.ASM15807v1.dna.nonchromosomal.fa.gz
!wget ftp://ftp.ensemblgenomes.org/pub/release-38/bacteria//fasta/bacteria_176_collection/_bacillus_aminovorans/dna/_bacillus_aminovorans.ASM164324v1.dna.nonchromosomal.fa.gz
!wget https://downloads.yeastgenome.org/sequence/S288C_reference/orf_dna/orf_genomic_1000_all.fasta.gz 
# Give yeast file a better name
!mv  orf_genomic_1000_all.fasta.gz yeast_genomic_1000_all.fa.gz 

You should be able to see these four files in your directory now.

In [ ]:
# See our four FASTA genome files
ls *.fa.gz

There are many techniques to identify which species you have sequenced. One approach would be to try to align each of our contigs to each reference in our database. However, that would be extremely computationally intensive. Instead here we will use a computationally light application called MASH, which calculates an estimate for the distance between our query sequence and each reference.  In this context, distance refers to how similar or different two genomes are.

# Mash for taxonomy classification
First we must form a sketch from each of our reference genomes. A sketch is a minified representation of a reference. This contains less information than a whole genome sequence, but still allows us to calculate the distance between each of the references and our query.

We have already downloaded the sequences for each sample species described above. Now we must build the sketch of each reference using the command sketch from the mash tool.

In [ ]:
# The -o option allows us to name the sketch output file.
!mash sketch /work/data/*.fa.gz -o reference

Now we have our database of references in one file, reference.msh.  We want to take our assembly from before and compare it to the reference database.

When we assembled our contigs using miniasm, it tried to piece together all of the individual strands of DNA sequenced by the Nanopore.  If more than one species was present or the assembler couldn't figure out where to place small pieces of the organism's genome, then the assembly will be made of multiple contigs.  If we sequenced human DNA, we would naturally get many contigs, one for each chromosome.  In practice, it's hard to piece together the whole genome so even within one chromosome some gaps may remain unfilled.

To build a sketch for each contig we must us the -i option of the sketch command. This option tells mash to treat each contig as its own seperate entity. Here we again use the -o option as described above.

In [ ]:
# Note: if there's no star on the left side of the code block ("In [*]:") then you know it's done running.
!mash sketch -i asm.fa -o contigs

Now we are prepared to compute the distances between each contig and each reference. This is accomplished using the dist command. The '>' signifies to take the output from the following program and create a file called distance.tab.

In [ ]:
!mash dist reference.msh contigs.msh > distance.tab

We can use the head command to look at the first five lines of this distance file. This file is organized into the following columns: [reference-ID, query-ID, distance, p-value, shared-hashes]

In [ ]:
!head distance.tab

# Using MetaGenomeMark to Describe Gene Information
- meta genome mark will take the fasta file of the assembled genome
- tell you what genes are in the genome


genomeOutputAssembled.fa - input file from Canu

In [ ]:
!pwd
!ls /work/MetaGeneMark_linux_64/mgm

## This will obtain a GFF file

In [ ]:
!gmhmmp -a -r -f G -d -m ../MetaGeneMark_linux_64/mgm/MetaGeneMark_v1.mod -o data/sequence.gff assembly.fa


### Using this gff file, we can learn what genes are in your sample!

A GFF file has the columns: 
- seqname - name of the chromosome or scaffold
- source - name of the program that generated this feature: GeneMark.hmm
- feature - name of Gene, Variation, or Similarity
- start - Start position of the feature, with sequence numbering starting at 1.
- end - End position of the feature, with sequence numbering starting at 1.
- score - A floating point value.
- strand - defined as + (forward) or - (reverse).
- frame - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..
- attribute - A semicolon-separated list of tag-value pairs, providing additional information about each feature.

### Quick look at GFF file

In [ ]:
!head -20 data/sequence.gff

### Goal: Getting FASTA files of all contigs listed 
This is your assembly file

In [ ]:
!head data/asm.fa

### The GFF file has unecessary headers that we don't need.
Run this sed command in unix to remove them

In [ ]:
!sed -i -e 1,9d data/sequence.gff

This is our current GFF file, followed how many contigs found

In [ ]:
!head -20 data/sequence.gff
!wc -l data/sequence.gff

### Run this to slice the FASTA sequences from the assembly from the GFF start and stop indecies

In [ ]:
# This python script will get the start and stop indexes from the GFF 
# and get FASTA sequences from the assembly 

import csv

nameOfContig = list()
startIndexList = list()
stopIndexList = list()
# get start and stop indexes in the GFF file
with open("data/sequence.gff") as tsv:
    for line in csv.reader(tsv, dialect="excel-tab"): #You can also use delimiter="\t" rather than giving a dialect.
        if len(line) > 1:
            nameOfContig.append(""+str(line[2:3][0])+str(line[3:4][0])+"-"+str(line[4:5][0]))
            startIndexList.append(line[3:4])
            stopIndexList.append(line[4:5])
startAndStopList = list(zip(nameOfContig,startIndexList,stopIndexList))

# Use BioPython to assemble output FASTA file
from Bio import SeqIO
sequences = list()
for record in SeqIO.parse("data/asm.fa", "fasta"):
    print("This is the header for your assembly fasta: "+record.id)
    for name,start,stop in startAndStopList :
        if start != [] and stop != [] :
            sequences.append(record.seq[int(start[0]):int(stop[0])])
fastaList = list(zip(nameOfContig, sequences))
with open("data/annotatedGene.fa", "w") as output_handle:
    for name, seq in fastaList:
        fasta_format_string = ">"+name+"\n%s\n" % seq
        output_handle.write(fasta_format_string)

# Get the largest FASTA sequence
maxFasta = max(fastaList, key=lambda x: len(x[1]))
fasta_format_string = ">"+str(maxFasta[0])+"\n%s\n" % str(maxFasta[1])
print(fasta_format_string)

# Blastn the largest sequence
from Bio.Blast import NCBIWWW
result_handle = NCBIWWW.qblast("blastx", "nr", str(maxFasta[1]))



### Annotated FASTA file is in /data
- This created a fasta of all the gffs, but theres alot!
- Copy the fasta sequence printed out from the last command and use the BLAST website to find hits to species!